subsys/settings/settings_file.c

```diff
@@ -0,0 +1,613 @@
+/*
+ * Copyright (c) 2018 Nordic Semiconductor ASA
+ * Copyright (c) 2015 Runtime Inc
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <string.h>
+#include <errno.h>
+
+#include <settings/settings.h>
+#include "settings_priv.h"
+
+struct settings_file_read_fn_arg {
+	void *stor_ctx;
+	off_t seek; /* offset of value within the file */
+};
+
+struct file_entry {
+	off_t off_name;   /* offset of the start of name in file */
+	size_t len_name;  /* name len including '=' */
+	off_t off_value;  /* offset of the start of value in file */
+	size_t len_value; /* value length */
+};
+
+static int settings_file_load(struct settings_store *cs);
+static int settings_file_save(struct settings_store *cs, const char *name,
+			      const char *value, size_t val_len);
+
+static struct settings_store_itf settings_file_itf = {
+	.csi_load = settings_file_load,
+	.csi_save = settings_file_save,
+};
+
+static ssize_t settings_file_read_fn(void *back_end, void *data, size_t len)
+{
+	struct settings_file_read_fn_arg *rd_fn_arg = back_end;
+	struct fs_file_t *file = rd_fn_arg->stor_ctx;
+	int rc;
+
+	rc = fs_seek(file, rd_fn_arg->seek, FS_SEEK_SET);
+	if (rc < 0) {
+		return rc;
+	}
+
+	rc = fs_read(file, data, len);
+	return rc;
+}
+/*
+ * Register a file to be a source of configuration.
+ */
+int settings_file_src(struct settings_file *cf)
+{
+	cf->cf_store.cs_itf = &settings_file_itf;
+	settings_src_register(&cf->cf_store);
+
+	return 0;
+}
+
+int settings_file_dst(struct settings_file *cf)
+{
+	cf->cf_store.cs_itf = &settings_file_itf;
+	settings_dst_register(&cf->cf_store);
+
+	return 0;
+}
+
+/*
+ * Called to advance to a next entry in a file
+ *
+ * Items are stored as: string - binary where
```

Comment:
Because it is much easier to construct a read function if it is stored as ```name=<len><value>```, and if encoding would be used (which makes sense for file backend) the storage would be ```name=<encoded value>```. The ```<len><value>``` construct is just a method of encoding.

Reply:
If it is more convince for you - that's ok.